File: src/main/java/uk/gov/ons/ssdc/caseprocessor/utils/CaseRefGenerator.java
package uk.gov.ons.ssdc.caseprocessor.utils;

import uk.gov.ons.ssdc.caseprocessor.utils.pseudorandom.PseudorandomNumberGenerator;

public class CaseRefGenerator {
  //  This gives a caseRef 9 long, plus 1 checkbit at the end = 10
  private static final int LOWEST_POSSIBLE_CASE_REF = 100000000;
  private static final int HIGHEST_POSSIBLE_CASE_REF = 999999999;

  private static final PseudorandomNumberGenerator PSEUDORANDOM_NUMBER_GENERATOR =
      new PseudorandomNumberGenerator(HIGHEST_POSSIBLE_CASE_REF - LOWEST_POSSIBLE_CASE_REF);

  public static long getCaseRef(int sequenceNumber, byte[] caserefgeneratorkey) {
    // DO NOT replace this with a random number generator - we must have zero collisions/duplicates
    int pseudorandomNumber =
        PSEUDORANDOM_NUMBER_GENERATOR.getPseudorandom(sequenceNumber, caserefgeneratorkey);
    int caserefWithoutCheckDigit = pseudorandomNumber + LOWEST_POSSIBLE_CASE_REF;
    return LuhnHelper.addCheckDigit(caserefWithoutCheckDigit);
  }
}

File: python spike/GCP PubSub/caseprocessor/case_ref_generator.py
"""
Just using random for now for simplicity
"""

from .pseudorandom.pseudo_random_number_generator import PseudorandomNumberGenerator
from .pseudorandom.luhn_utility import add_check_digit

# This gives a caseRef 9 long, plus 1 checkbit at the end = 10
LOWEST_POSSIBLE_CASE_REF = 100000000
HIGHEST_POSSIBLE_CASE_REF = 999999999

PSEUDORANDOM_NUMBER_GENERATOR = PseudorandomNumberGenerator(HIGHEST_POSSIBLE_CASE_REF - LOWEST_POSSIBLE_CASE_REF)


def get_case_ref(sequence_number: int, case_ref_generator_key: bytes):
    pseudo_random_number = PSEUDORANDOM_NUMBER_GENERATOR.get_pseudorandom(sequence_number,
                                                                          case_ref_generator_key)

    case_ref_without_check_digit = pseudo_random_number + LOWEST_POSSIBLE_CASE_REF
    return add_check_digit(case_ref_without_check_digit)
